[distributed] use timeout in connect function to prevent againstinfinite loop

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26364 [distributed] use timeout in connect function to prevent againstinfinite loop**
infinite loop**
infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

rohan-varma committed:
[pytorch][distributed] use timeout in connect function to prevent against
infinite loop

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Connecting to TCP store timed out.
Aborted (core dumped)
```

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -231,6 +236,23 @@ int connect(
+          if (elapsedMilliseconds > timeout.count()) {
+            throw std::runtime_error("Connecting to TCP store timed out.");
+          }
+        }
```
No need to take `time_since_epoch`. You can use the `std::chrono::time_point` and `std::chrono::duration` directly without bothering with the unit (the duration class takes care of the comparison correctly if the unit is different):

```c++
const auto start = std::chrono::high_resolution_clock::now();

// ...

const auto elapsed = std::chrono::high_resolution_clock::now() - start;
if (elapsed > timeout) {
  // ...
}
```

pietern commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -231,6 +236,23 @@ int connect(
+        /*
+        if a timeout is specified, check time elapsed to see if we need to
+        timeout. A timeout is specified if timeout != kNoTimeout.
+        */
```
Comment style is `//` pretty much everywhere. There are some blocks with `/* xyz */` but those are typically limited to large chunks of documentation outside of a function. Here you can stick to `//`.

pietern commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -231,6 +236,23 @@ int connect(
+
+          auto elapsedMilliseconds = nowMilliseconds - startMilliseconds;
+          if (elapsedMilliseconds > timeout.count()) {
+            throw std::runtime_error("Connecting to TCP store timed out.");
```
There is a different timeout error thrown above, where `poll(2)` returns with a timeout. Please consolidate so there is only one message for connect timeout.

rohan-varma replied:
Sounds good. The errors are slightly different though, one is because `poll(2)` timed out on a single attempt, while this error can happen because we keep getting `ECONNREFUSED` or `ECONNRESET` after trying many times. Is it worth having different error messages to help with debugging?

pietern replied:
I think it's fine to be the same and that multiple add to the confusion here.

pietern reviewed (commented):
The call to `poll` uses `timeout.count()` as its timeout. It should use the time remaining instead, because then it accounts for the time already spent. I'm thinking this should do the trick (I have the `std::chrono` docs in front of me now anyway): 

```c++
const auto elapsed = std::chrono::high_resolution_clock::now() - start;
const auto remaining = std::chrono::duration_cast<std::chrono::milliseconds>(timeout) -
  std::chrono::duration_cast<std::chrono::milliseconds>(elapsed);
const auto remainingMillis = std::max(0, left.count());
```

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Connecting to TCP store timed out.
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Connecting to TCP store timed out.
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma commented:
> The call to `poll` uses `timeout.count()` as its timeout. It should use the time remaining instead, because then it accounts for the time already spent.

Sounds good, we should be careful to respect if `timeout = kNoTimeout` though. I'll condition the code on this, because if we don't, then if `timeout = kNoTimeout` (which is aliased to -1) we will actually pass in a timeout to `0` to poll, instead of -1, changing its semantics.

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**
infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -185,7 +186,16 @@ int connect(
+        const auto remaining =
+            std::chrono::duration_cast<std::chrono::milliseconds>(timeout) -
+            std::chrono::duration_cast<std::chrono::milliseconds>(elapsed);
+        pollTimeout = std::max(0, (int)remaining.count());
```
have to cast this to int, or else the macOS compiler throws errors.

pritamdamania87 replied:
You should cast to int64_t (or use 0LL in std::max()), also use `static_cast<int64_t>` instead of C style casts.

rohan-varma commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -231,6 +241,16 @@ int connect(
+          const auto elapsed =
+              std::chrono::high_resolution_clock::now() - start;
+          if (elapsed > timeout) {
+            throw std::runtime_error("connect() timed out");
```
The error msg now reads the same as the one we throw when `poll` times out. Should we make this a const string somewhere?

pietern replied:
Yes, good idea. We like to use `kSomeName` for globals, but this one can be kept local only in an anonymous namespace (equivalent of static global), since it's only used in this file.

rohan-varma commented:
@pytorchbot retest this please

pietern commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -185,7 +186,16 @@ int connect(
       pfd.events = POLLOUT;
 
-      int numReady = ::poll(&pfd, 1, timeout.count());
+      int pollTimeout = timeout.count();
```
It reads a bit weird that you first set `pollTimeout` and then check the other variable. Could you instead assign `pollTimeout = -1` to be clear that this is the default? This makes it clear(er) IMO that it'll have a negative value if `timeout == kNoTimeout`. As is, we could redefine `kNoTimeout` to be 0 and not be aware of any harm.

pietern reviewed (approved):
Minor comment regarding robustness. Feel free to merge once that's taken care of. Thanks, @rohan-varma!

pritamdamania87 commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -231,6 +241,16 @@ int connect(
+        }
         std::this_thread::sleep_for(std::chrono::seconds(1));
         anyRefused = false;
         anyReset = false;
```
@pietern Why do we continue retrying here if we have already exhausted all addresses?

pietern replied:
The process that is supposed to start the TCP initialization daemon may not yet have started and then the host will reply with `ECONNREFUSED`. Somewhat similar, if the daemon has started, but 1000 processes connect at the same time (since they are typically all started at the same time), they may exhaust the listen backlog, and if full, the host replies with a RST packet (hence `ECONNRESET`). In both cases, we should retry until the timeout hits.

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[pytorch][distributed] use timeout in connect function to prevent against
infinite loop"

infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] use timeout in connect function to prevent againstinfinite loop"

infinite loop**
infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/Utils.cpp:
```diff
@@ -185,12 +187,22 @@ int connect(
+            std::chrono::duration_cast<std::chrono::milliseconds>(timeout) -
+            std::chrono::duration_cast<std::chrono::milliseconds>(elapsed);
+        pollTimeout = std::max(
+            static_cast<int64_t>(0), static_cast<int64_t>(remaining.count()));
```
This is ugly, but the other variants (std::max(0L, ...), std::max(0LL, ...) all caused compiler errors

rohan-varma committed:
Update on "[distributed] use timeout in connect function to prevent againstinfinite loop"

infinite loop**
infinite loop**
infinite loop**

Per https://github.com/pytorch/pytorch/issues/25769, we sometimes get
an infinite loop when `TCPStore` calls `tcputil::connect`, and the server
continually returns `ECONNRESET` or `ECONNREFUSED`. If a proper timeout is passed
in, we guard against this by throwing an exception once the timeout has passed.

Testing: Tested with modifying `TCPStore` to connect to an invalid port, thus getting
`ECONNREFUSED`. If a valid timeout is passed in, the function correctly throws an
exception. Steps below:
1) in TCPStore.cpp's constructor, replace the `connect` call with this line:
 `storeSocket_ = tcputil::connect(tcpStoreAddr_, 1, true, std::chrono::milliseconds(3000));`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Expected output:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  connect() timed out
Aborted (core dumped)
```

We should probably add tests around this, but there are some issues. Namely, the `connect` call in `TCPStore.cpp` does not pass in a `timeout` param at all right now (https://github.com/pytorch/pytorch/blob/929764ac2ad0d7c26e3defa2d4c4d39f4933a7ef/torch/lib/c10d/TCPStore.cpp#L290). We would need to change it to pass this param in, and then call `serverTCPStore.setTimeout()` in the test.

Differential Revision: [D17430164](https://our.internmc.facebook.com/intern/diff/D17430164/)

[ghstack-poisoned]
